### PR TITLE
[v2] Add a new spawn namespace to contain all spawn logic

### PR DESCRIPTION
**Pull request by neg**
Hi Richard,

This split is something I have pondered for a while as I carry a few different "shell" implementations out-of-tree for things like picocom, ssh, minicom and a few other real odd things. I have always had wrappers like this to make it work.

```
package provide myssh 1.0

package require 9pm

namespace eval myssh {
    proc open {alias node} {
        if {![::9pm::shell::open $alias]} {
            return FALSE
        }

        ::9pm::ssh::connect $node

        return TRUE
    }

    proc close {alias} {
        return [::9pm::shell::close $alias]
    }
}
```

I always felt it it was a bit of a hack and it don't work with shell::push and shell:pop, which for my use-cases have been OK. But today I needed that and instead of trying to hack that into an out-of-tree wrapper I will attempt to solve it upstream as it can be useful for someone else.

Compared to v1 I have implemented the namespaces topology I briefly touched in v1 cover-letter.

-     ::9pm::spawn - Holds the base spawn manipulation logic
-     ::9pm::spawn::shell - Holds the ::9pm::shell logic
-     ::9pm::spawn::ssh - Holds the new ssh spawn logic

I have created this topology by keeping backwards compatibility with the old shell implementation by turning it into a wrapper for the later.

### CODE REVIEW

File: spawn.tcl [hunk 1]
```diff
@@ -134,4 +134,57 @@ namespace eval ::9pm::spawn {
...
+
+            set pid [spawn "ssh" "$hostname"]
+            if {$pid == 0} {
+                ::9pm::fatal ::9pm::output::fail "Failled to connect to $hostname"
+            }
+
+            # Wait for prompt
+            expect {
+                $prompt {
+                    ::9pm::output::info "Connected to \"$hostname\""
```
Comment (neg) [inline spawn.tcl:152]: This should be ::9pm::output::debug

**Comment by rical** [comment]
We have very different use-cases. The idea of shells and locations was done intentionally to avoid having this spawn logic. However, I don't know your use-case and I trust you have a valid reason for this change.

What if we did something like this.

Old fashion way, where the first spawn is a shell.
```
9pm::shell::open "usa-console"
9pm::ssh::connect $usa
9pm::console::connect $device
9pm::cmd::execute "reboot"
...
9pm::shell::open "usa-console"
9pm::cmd::execute "foobar"
```
New way, where the first spawn is the ssh process.
```
9pm::spawn::ssh::connect "usa-console" $usa
9pm::console::connect $device
9pm::cmd::execute "reboot"
...
9pm::spawn::open "usa-console"
9pm::cmd::execute "foobar"
```
Where `9pm::spawn::ssh::connect` does `9pm::ssh::create` or  `9pm::ssh:spawn_and_connect` or similar.
The idea is to move the ssh and shell logic out of the spawn context and only keep the new "create proc callbacks".

If we take ssh as an example we might have something like this (pseudocode):

```
9pm::spawn::ssh::connect {alias data} {
    9pm::ssh::spawn_and_connect $alias $data
}

9pm::ssh::spawn_and_connect {alias data} {
  spawn ssh .....
  9pm::spawn::add $alias $pid ....
  9pm::ssh::int::establish_connection $data
}

9pm::ssh::connect {alias data} {
  send ssh .....
  9pm::ssh::int::establish_connection $data
}
```
Let me know what you think, does this make sense to you?

**Comment by neg** [comment]
Hello Richard, 

Thanks for your feedback.

> We have very different use-cases. The idea of shells and locations was 
> done intentionally to avoid having this spawn logic. However, I don't 
> know your use-case and I trust you have a valid reason for this 
> change.

I agree our use-cases differ a bit. I think we have argued over this in 
the past :-) The important change here for me is the first two patches.  
Patch 3/3 is just to demonstrate that the new ::9pm::spawn:: base can be 
used to implement both a shell in ::9pm::spawn::shell and a ssh 
connecting in ::9pm::spawn::ssh.

I can do ::9pm::spawn::ssh  one out-of-tree if you feel it's confusing 
to have both ::9pm::spawn::ssh which spawns an ssh connection directly 
and the ::9pm::ssh:: helper to execute a ssh command in an already open 
spawn.

> 
> What if we did something like this.
> 
> Old fashion way, where the first spawn is a shell.
> 
> ```
> 9pm::shell::open "usa-console"
> 9pm::ssh::connect $usa
> 9pm::console::connect $device
> 9pm::cmd::execute "reboot"
> ...
> 9pm::shell::open "usa-console"
> 9pm::cmd::execute "foobar"
> ```
> 
> New way, where the first spawn is the ssh process.
> 
> ```
> 9pm::spawn::ssh::connect "usa-console" $usa
> 9pm::console::connect $device
> 9pm::cmd::execute "reboot"
> ...
> 9pm::spawn::open "usa-console"
> 9pm::cmd::execute "foobar"
> ```

This is what this series is proposing apart from the name
s/9pm::spawn::ssh::connect/9pm::spawn::ssh::open/, or have a missed a 
use-case somewhere? The old fashion way still works exactly as before 
and the new way works exactly as you propose above, apart from the name.

I like open over connect in this setting, but if you feel strongly about 
it I can live with connect.

> 
> Where `9pm::spawn::ssh::connect` does `9pm::ssh::create` or 
> `9pm::ssh:spawn_and_connect` or similar. The idea is to move the ssh 
> and shell logic out of the spawn context and only keep the new "create 
> proc callbacks".
> 
> If we take ssh as an example we might have something like this (pseudocode):
> 
> ```
> 9pm::spawn::ssh::connect {alias data} {
>     9pm::ssh::spawn_and_connect $alias $data
> }
> 
> 9pm::ssh::spawn_and_connect {alias data} {
>   spawn ssh .....
>   9pm::spawn::add $alias $pid ....
>   9pm::ssh::int::establish_connection $data
> }
> 
> 9pm::ssh::connect {alias data} {
>   send ssh .....
>   9pm::ssh::int::establish_connection $data
> }
> ```
> 
> Let me know what you think, does this make sense to you?

I agree, this is the weak point in this submission and I always intended 
to refactor ssh.tcl to expose it's logic in a more modular way so it 
could be reused by spawn.tcl. But I did not want to touch ssh.tcl as a 
first step to reduce the number of changes.

I can add a patch as a part of this series or as follow up work, 
whatever works best for you. If possible I would like to avoid delaying 
merging of 1/3 and 2/3 as they block my progress elsewhere.

**Comment by neg** [comment]
I will post a v3.

**Pull request closed without merging**